core/storage/base_model/gae_models.py

```diff
@@ -214,6 +214,19 @@ def get_field_names_for_takeout(cls):
         """
         return {}
 
+    @classmethod
+    def get_marked_as_deleted(cls, entity_id):
```

Comment:
Isn't this effectively the same as just calling ``get()``? It may be worthwhile in the future to fork ``get()`` into two versions: one for strict & non-strict (rather than using a boolean), but having the boolean version & a non-boolean version actually muddies the API contract & may make it more confusing for developers to decide on which to use.

Reply:
The difference between `get` and `get_marked_as_deleted` is that `get` returns `None` for the models with field `deleted` set to `True`. Strict and non-strict versions work a bit differently since they decide whether errors should be returned or not.

Reply:
This name doesn't seem to convey exactly the right thing, then. It implies that the returned model is deleted, or should be marked as deleted, not that it specifically can be. I'm actually now starting to get confused from a reading perspective the differences between multiple gets & strict/non-strict. Can we consolidate the API to a single get() method with all needed controls, or a few consistent get methods that enumerate all the cases we care about? For example:

Option 1:
```python
def get(cls, entity_id, deletion_retrieval_policy)
    # deletion_retrieval_policy is one of three enum values: RETURN_DELETED, FILTER_DELETED, RAISE_ON_DELETED.
    ....
```

Option 2:
```python
def get_entity(entity_id):
    # Can be deleted; doesn't matter.

def get_existing_entity(entity_id):
    # Must be existing, so if it's deleted raise exception.

def get_entity_if_existing(entity-id):
     # Return entity if it exists, otherwise None.
```

The names for option 2 could use some iteration.

I think my main confusion is: the current name doesn't sound like what's actually happening, and as a user I still don't know when to use one get/property versus the other.

Reply:
I might be repeating a few things that Ben is saying but from my understanding.

I think what we have now is situation like this
<table>
<tr>
    <td> 
	<td> get only models that are not marked as deleted
	<td> get even models that are marked as deleted
<tr>
    <td> raise when the model is not found
	<td> get(id, strict=True)
	<td> we do not implement this on storage layer
<tr>
    <td> return None when the model is not found
	<td> get(id, strict=False)
	<td> get_by_id(id) (in this PR replaced by get_marked_as_deleted(id))
</table>

So I think we should cover all the possible cases, since in many places where we now use `get_by_id` we would benefit from the option to start using the __strict__ mode.

Option 1: Have a separate method for all four cases.
Option 2: Have one `get` method with two additional args, one for the strictness and the second for including the models marked as deleted.
Option 3: Have one `get` method with one additional arg, that arg would receive one of four values, the values would approx. be `strict_include_deleted`, `non_strict_include_deleted`, `strict_exclude_deleted`, `non_strict_exclude_deleted`.

Reply:
https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3 gets into this a bit (& there are other examples), but booleans introduce avoidable complexity that is paid at callsites. If we can come up with named methods that perform the exact contracts that we're looking for, I think that's always going to be the ideal solution since it eliminates any potential ambiguity at the callsite. This all being said, named arguments in Python do mitigate this some so I think option 2 could be reasonable if we have good names for the parameters ('strict' probably should be changed in this case since it's rather vague about what specifically is being strict or what that even means).

I did have some difficulty in coming up with different names, but maybe approaching it with all 4 cases in mind will be easier. If we can't come up with clear names, then option 2 with clear parameters is approach I would vote for.

Reply:
I also think that at least the name should be changed. I actually thought `get_marked_as_deleted` meant it only returned entities that were marked as deleted, which is not the case.

Reply:
https://docs.google.com/document/d/1YvaJSwNBvfsyPCsCw72KsqgXq4RDZ943EwPmwmd0vn4/edit?usp=sharing created a quick brainstorming doc for this

Reply:
Following up directly on the doc.